Feature: add bulk library actions

rhell4 opened the pull request:
This includes:
- Bulk update libraries to latest versions
- Bulk upgrade content to latest library version (button will be disabled if there is no content that needs upgrading)
- Bulk export libraries (for importing to another site)
- Bulk import is already a feature using the existing "Upload Libraries"

You'll find the 3 new buttons on the /mod/hvp/library_list.php page
![2024-08-28_11-20](https://github.com/user-attachments/assets/2977d276-8228-4fb7-b6fc-37c92a5a39c2)

I tested these new options using the following 2 H5Ps in this zip; [test-h5p.zip](https://github.com/user-attachments/files/16773544/test-h5p.zip)
Both H5Ps are using older libraries, so you should be able to test updating the libraries and the content.

When updating the libraries ensure the content type cache is up to date using the "Update content type cache" at the top of the page because this is required to check if a library has a newer version.

For the bulk export/import, I ran the export on the site I was testing on and then imported into a clean new site I had.

I've also fixed some PHP warnings/deprecations that were filling up the screen/logs when updating or uploading libraries.

Peterburnett commented on export_libraries.php:
```diff
@@ -0,0 +1,172 @@
+require_once('locallib.php');
+
+// No guest autologin.
+require_login(0, false);
```
is this require login actually required? admin_externalpage_setup already does this for you

rhell4 replied:
You're right this isn't needed, have removed it

Peterburnett commented on export_libraries.php:
```diff
@@ -0,0 +1,172 @@
+
+$libraries = $core->h5pF->loadLibraries();
+$tmppath = $exporter->h5pC->fs->getTmpPath();
+mkdir($tmppath, 0777, true);
```
Should we wrap this inside a make_request dir or something so that its always cleaned up irregardless of H5P?

rhell4 replied:
Good idea, I've switched all the H5P tmp path stuff over to use make_request_directory.

Peterburnett commented on update_libraries.php:
```diff
@@ -0,0 +1,143 @@
+require_once('locallib.php');
+
+// No guest autologin.
+require_login(0, false);
```
This shouldnt be necessary

rhell4 replied:
Removed

Peterburnett commented on classes/file_storage.php:
```diff
@@ -544,6 +544,7 @@ private static function readFileTree($source, $options, $archive = null, $relati
         if (empty($archive) && $exportzip) {
             $archive = new \ZipArchive();
             $path = tempnam(get_request_storage_directory(),'libdir');
+            unlink($path);
```
instead of unlinking, it looks like the line below is buggy, and should be | instead of || to do the bitwise OR for the flag. That might just supress the message.

rhell4 replied:
You're right it did turn out to be an issue with that, that's corrected now

Peterburnett commented on upgrade_all_content.php:
```diff
@@ -0,0 +1,134 @@
+                        'to' => $library->title . ' (' . $tover . ')',
+                        'count' => $numcontents,
+                    ];
+                    $settings['libraryInfo'][] = [
```
Can we move out the non-dynamic stuff then just copy and modify the non static fields in the loop? Saves a ton of string loads, and if there are a lot of libs, that might overrun the string cache, it only has room for 2000 strings in APC by default

rhell4 replied:
Good idea, I've moved all of the non dynamic stuff into it's own separate array and adjusted the js to use that

Peterburnett commented on classes/editor_ajax.php:
```diff
@@ -65,7 +65,8 @@ public function getLatestLibraryVersions() {
             SELECT hl4.id, hl4.machine_name, hl4.title, hl4.major_version,
-                hl4.minor_version, hl4.patch_version, hl4.has_icon, hl4.restricted
+                hl4.minor_version, hl4.patch_version, hl4.has_icon, hl4.restricted,
+                0 AS patch_version_in_folder_name
```
What does this do, im not seeing this field name referenced anywhere in the PR?

rhell4 replied:
This is to fix up a key doesn't exist warning here in the editor submodule https://github.com/h5p/h5p-editor-php-library/blob/53dc7bdc57b17f5a0d55a8871a36d1b1bd631e70/h5peditor.class.php#L633

rhell4 committed:
Fix: php warnings and deprecations

rhell4 committed:
Feature: add bulk libray actions

This includes:
- Bulk update libraries to latest versions
- Bulk upgrade content to latest library version
- Bulk export libraries (for importing to another site)
- Bulk import is already a feature using the existing "Upload Libraries"

Peterburnett commented:
Thanks Rossco, this is looking great!. For any future folks, CI is cooked due a ton of lint errors from the main repo, but the tests are passing.
